8282270: java/awt/Robot Screen Capture tests fail after 8280861

mkartashev opened the pull request:
The two tests `ScreenCaptureGtkTest.java` and `HiDPIRobotScreenCaptureTest.java` under `java/awt/Robot/HiDPIScreenCapture` started to intermittently fail under Windows and Linux after the [recent changes](https://github.com/openjdk/jdk/commit/cc7cf81256ed4d74493472017b1c4df20fa2208a) made exclusively for the Linux `Robot` implementation.

`HiDPIRobotScreenCaptureTest.java`
Since the Windows failures cannot possibly have their origin in the fix made - the latter being Linux-only - the test apparently either exposes an existing bug in the Windows `Robot` or is bumping against a peculiarity of that platform. The test is reverted back to its original form that didn't fail.

`ScreenCaptureGtkTest.java`
The coordinates in the log `(83, 78)` of a failure are higher up than the test suggests `(83, 97)`. I've seen similar failures on Ubuntu 20.04 when the coordinates were set to `(0, 0)`. The color then picked matched the color of the bar drawn at the top of the screen. I believe it's best to place the test pixels towards the center of the window to avoid desktop elements interference.

The other possible reason for intermittent failures are window coordinates. Suppose that the test is executing with `sun.java2d.uiScale` set to 2. This effectively means that it can only pick colors of pixels at even coordinates (in the absolute desktop space). So if the window is placed at, say, `(201, 201)`, `Robot` can only pick the color at either `(200, 200)` or `(202, 202)`. Since the test only makes sense if it is pixel-accurate, I removed all `@run`s with `sun.java2d.uiScale` other than 1. This way window placement will not cause a failure.

This was tested on Ubuntu 20.04 and 18.04 with desktop scaling set to 100%, 200%, and 300%.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282270](https://bugs.openjdk.java.net/browse/JDK-8282270): java/awt/Robot Screen Capture tests fail after 8280861


### Reviewers
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7613/head:pull/7613` \
`$ git checkout pull/7613`

Update a local copy of the PR: \
`$ git checkout pull/7613` \
`$ git pull https://git.openjdk.java.net/jdk pull/7613/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7613`

View PR using the GUI difftool: \
`$ git pr show -t 7613`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7613.diff">https://git.openjdk.java.net/jdk/pull/7613.diff</a>

</details>

mkartashev commented:
@prrace In the bug you [suggested](https://bugs.openjdk.java.net/browse/JDK-8282270?focusedCommentId=14478682&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14478682) that the fix shall be put through CI testing a couple of times. Unfortunately, I don't know how to do that and would appreciate some guidance in that direction. Thank you!

mrserb commented:
If we do not know the root cause of the HiDPIRobotScreenCaptureTest.java failure why we roll that change back since it could be a product bug?

mkartashev commented:
> I submitted a job for testing the changes.

Thanks a lot!

mkartashev commented:
> If we do not know the root cause of the HiDPIRobotScreenCaptureTest.java failure why we roll that change back since it could be a product bug?

@mrserb I can quote the source of the test and file a Windows-specific bug saying that "this code intermittently fails" if that helps.

aivanov-jdk commented:
> > I submitted a job for testing the changes.
> 
> Thanks a lot!

I got 1 failure out of 15 on Linux for `java/awt/Robot/HiDPIScreenCapture/HiDPIRobotScreenCaptureTest.java`. The screenshot of the entire screen has no open windows: just plain desktop background. It was probably captured after the test was shutdown.

The test doesn't save the captured image on failure, does it?

It's crucial to save the captured images, otherwise we're blind and left to guess what Robot captured and why it didn't match.

mkartashev commented:
Updated the failing test by increasing the delay between when the frame is shown and image capture. Also, added code to save an image in case of failure. The change was verified on Ubuntu 20.04 and Windows 10.

@aivanov-jdk  Could you please re-run the test? Thanks once more!

prrace commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/ScreenCaptureGtkTest.java:
```diff
@@ -40,27 +40,26 @@
  * @run main/othervm -Djdk.gtk.version=3 -Dsun.java2d.uiScale=1 ScreenCaptureGtkTest
- * @run main/othervm -Djdk.gtk.version=3 -Dsun.java2d.uiScale=2 ScreenCaptureGtkTest
- * @run main/othervm -Djdk.gtk.version=3 -Dsun.java2d.uiScale=3 ScreenCaptureGtkTest
  */
```
So having removed all of these, it isn't testing hi-dpi any more, is it ? Even if running on a hi-dpi screen.
Would it be better to let it just use whatever it wants for default ? Meaning if it happens to run on a hi-dpi
screen we want to know if that works don't we ?

mkartashev replied:
> So having removed all of these, it isn't testing hi-dpi any more, is it ?

True. But there still is `ScreenCaptureTest.java` that will continue testing hi-dpi. 

Keeping `ScreenCaptureGtkTest.java` still makes sense because it verifies that `getPixelColor()` works precisely when neighboring pixels are of a different color; also, it shuffles the GTK versions.

Unfortunately, I see no way to fully and reliably test the changes for 8280861 with scaling above 1. I am forced to choose between two evils: 
1. either I set the scale above 1 and pick color inside a large area to be sure that I don't pick neighboring pixels because of an unfortunate window placement
1. or I set the scale to 1 and test colors of pixels sitting next to one another.

(1) was tested by the existing tests, (2) is tested by `ScreenCaptureGtkTest.java`.

mrserb replied:
you can check both cases, if the scale=1 then use high precision.
if the scale > 1 then you can use low precision -> if the scale is equal to 2 then shift by two/four pixels or something like that?

mkartashev replied:
> if the scale > 1 then you can use low precision -> if the scale is equal to 2 then shift by two/four pixels or something like that?

@mrserb I suppose I could do that, but then it wouldn't be much different from the existing tests that verify color inside a sufficiently large area. I can make this area smaller, but this still won't help to find a rounding error in `gtk3_get_drawable_data()`.

aivanov-jdk replied:
You mentioned uiScale=3 doesn't work correctly. Does it not work with uiScale=2 either?

Is it proven the test can't get the colored pixels precisely with uiScale=2?

mkartashev replied:
> You mentioned uiScale=3 doesn't work correctly. Does it not work with uiScale=2 either?
> 
> Is it proven the test can't get the colored pixels precisely with uiScale=2?

Yes to both.

aivanov-jdk commented:
> @aivanov-jdk Could you please re-run the test? Thanks once more!

Sorry for my delayed reply.

One failure reported:
```
command: main -Dsun.java2d.win.uiScaleX=3 -Dsun.java2d.win.uiScaleY=2 HiDPIRobotScreenCaptureTest
...
----------System.out:(3/218)----------
Creating screen capture of java.awt.Rectangle[x=83,y=97,width=400,height=300]
Checking (100, 75) to have color java.awt.Color[r=0,g=255,b=0]
... Mismatch: found java.awt.Color[r=135,g=8,b=69] instead. Check image.png.
----------System.err:(12/723)----------
java.lang.RuntimeException: Wrong image color!
	at HiDPIRobotScreenCaptureTest.checkRectColor(HiDPIRobotScreenCaptureTest.java:125)
	at HiDPIRobotScreenCaptureTest.main(HiDPIRobotScreenCaptureTest.java:108)
...
```
![The captured image.png](https://user-images.githubusercontent.com/70774172/157277072-febab0d6-bef3-43df-a98f-8b6b6eea60ff.png)

The checked point (x, y) = (100, 75) is 10 pixels into the green square but it's 55 pixels above the square.

mkartashev commented:
@aivanov-jdk Thanks for re-running the test. Where does this failure come from - Linux or Windows? The saved image's size is correct (400x300 px), but I can't explain the location of the painted area inside it and why it is so small...

aivanov-jdk commented:
> @aivanov-jdk Thanks for re-running the test. Where does this failure come from - Linux or Windows? The saved image's size is correct (400x300 px), but I can't explain the location of the painted area inside it and why it is so small...

It's Linux, Ubuntu 20.10. Neither can I.

I checked the history and the test failed on different hosts, so it could be an intermittent bug in JDK itself.

~~Shouldn't the image be larger than 400×300? The size of the frame is 400×300 in user's space, the test is running with uiScaleX=3 and uiScaleY=2, therefore the size of the frame should be 1200×600. In some cases, such a size may not fit on the screen.~~ I see these `sun.java2d.win.uiScale{X,Y}` affect Windows only. On Ubuntu, the test runs according to the system settings.

aivanov-jdk commented:
> Shouldn't the image be larger than 400×300? The size of the frame is 400×300 in user's space, the test is running with uiScaleX=3 and uiScaleY=2, therefore the size of the frame should be 1200×600. In some cases, such a size may not fit on the screen.

This is what happened on Windows:
![Captured image on Windows](https://user-images.githubusercontent.com/70774172/157427018-4b7acdb9-145a-426c-888d-8d4be33b0fc0.png)

The frame didn't fit on the screen. According to the screenshot of the entire screen, the resolution on the host was 1024×768. It can't fit the frame which is larger than screen 1200×600, especially when it is positioned at 83*3=248, 97*2=194.

This failure has nothing to do with [JDK-8280861](https://bugs.openjdk.java.net/browse/JDK-8280861), it's just a coincidence.

mkartashev commented:
@aivanov-jdk I added image-saving code to the Linux test `ScreenCaptureGtkTest.java`. Could you please re-run it?
As before, it passes my test runs on Ubuntu 20.04.

aivanov-jdk commented:
> @aivanov-jdk I added image-saving code to the Linux test `ScreenCaptureGtkTest.java`. Could you please re-run it? As before, it passes my test runs on Ubuntu 20.04.

I added [the code myself](https://github.com/aivanov-jdk/jdk/commits/8282270-screenCapture) and it's scheduled to run on the host where the test failed. I started my branched from [JDK-8280861](https://bugs.openjdk.java.net/browse/JDK-8280861). Let's see what went wrong.

aivanov-jdk commented:
Does it make sense to run all the checks before throwing the exception?

I mean run all the checks that the test performs and accumulate failures if any. If the list of failures isn't empty, throw an exception to fail the entire test.

mkartashev commented:
> Does it make sense to run all the checks before throwing the exception?
> 
> I mean run all the checks that the test performs and accumulate failures if any. If the list of failures isn't empty, throw an exception to fail the entire test.

That might turn out to be useful for debugging a failure, I think.

aivanov-jdk commented:
ScreenCaptureGtkTest failed when it was run with `-Djdk.gtk.version=2 -Dsun.java2d.uiScale=3` options:

```
----------messages:(5/317)----------
command: main -Djdk.gtk.version=2 -Dsun.java2d.uiScale=3 ScreenCaptureGtkTest
...
----------System.out:(1/128)----------
Checking color at 83, 78 to be equal to java.awt.Color[r=0,g=255,b=0]...
Mismatch: found java.awt.Color[r=54,g=52,b=58] instead
----------System.err:(13/786)----------
Gtk-Message: 20:23:41.357: Failed to load module "canberra-gtk-module"
java.lang.RuntimeException: Wrong screen pixel color
	at ScreenCaptureGtkTest.checkPixelColors(ScreenCaptureGtkTest.java:127)
	at ScreenCaptureGtkTest.main(ScreenCaptureGtkTest.java:101)
```

The regular screenshot shows nothing: the image is filled with green with three pixel dots.

However, multi-resolution screenshot reveals something (the image is zoomed):
![MRI screenshot zoomed](https://user-images.githubusercontent.com/70774172/157537522-2d61eda5-d499-4a23-8a28-083c19bf4079.png)

The first pixel line above is part of the background image. The robot seems to pick it up it instead of the correct pixel.

The host where the test fails is Ubuntu 18.04.

mkartashev commented:
The `ScreenCaptureGtkTest` is not supposed to be run with `-Dsun.java2d.uiScale=3` anymore as we've discovered the problem with precise window positioning. Let's try this test from [my branch](https://github.com/mkartashev/jdk/tree/JDK-8282270).

aivanov-jdk commented:
> The `ScreenCaptureGtkTest` is not supposed to be run with `-Dsun.java2d.uiScale=3` anymore as we've discovered the problem with precise window positioning. Let's try this test from [my branch](https://github.com/mkartashev/jdk/tree/JDK-8282270).

I'll run them. However, the test passes successfully on the same host when the frame is smaller: 100×100 and 50×20. So it looks in both cases the reason for failure was that the frame was too large.

mkartashev commented:
The initial reason for `ScreenCaptureGtkTest` failure was that on Linux the window was *sometimes* positioned such that some of the pixels targeted by `Robot` couldn't be pointed to by the underlying GTK API. So these failures were intermittent and out of control of the test itself; change of the windows' size could've simply changed the window manager's logic for position calculation and thus accidentally made the test pass.

aivanov-jdk commented:
> I recognise that positioning is another issue but I've never seen a window manager position a window so that it is partially off-screen unless it cannot fit !

Windows allows positioning a window so that it's partially or completely off screen. It has happened to me a few times that an app displayed its window off screen when display arrangement changed from the previous start of the app. I had to use the window context menu to move it.

aivanov-jdk commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/HiDPIRobotScreenCaptureTest.java:
```diff
@@ -115,38 +105,26 @@ public void paint(Graphics g) {
+        checkRectColor(image, w / 4, h / 4, COLORS[0]);
+        checkRectColor(image, 3 * w / 4, h / 4, COLORS[1]);
+        checkRectColor(image, w / 4, 3 * h / 4, COLORS[2]);
+        checkRectColor(image, 3 * w / 4, 3 * h / 4, COLORS[3]);
```
As my testing showed, this test was stable enough with the changes introduced by [JDK-8282270](https://bugs.openjdk.java.net/browse/JDK-8282270), #7425, with the reduced size of the frame.

mkartashev replied:
I can revert the changes for this one test, but will do that in a separate commit so it's easier to revert in case there are troubles.

aivanov-jdk commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/ScreenCaptureGtkTest.java:
```diff
@@ -40,27 +46,26 @@
         // by the system because that area's occupied by the global
         // menu bar and such.
-        frame.setBounds(83, 97, 400, 300);
+        frame.setBounds(89, 99, 100, 100);
```
99 is not a prime number. Is it intentional?

mkartashev replied:
Now that there's an offset within the frame, the numbers should take that into account: (50+89)==139 and (50+99)==149 and these are prime numbers.

aivanov-jdk commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/ScreenCaptureGtkTest.java:
```diff
@@ -76,7 +81,7 @@ public void paint(Graphics g) {
                 for (int i = 1; i < 4; i++) {
                     g.setColor(COLORS[i]);
-                    g.fillRect(i, 0, 1, 1);
+                    g.fillRect(leftOffset + i, topOffset, 1, 1);
```
May I suggest using `COLORS.length` instead of 4 in the for-loops for painting and checking?

mkartashev replied:
Sure. Will fix that.

aivanov-jdk commented:
> > Does it make sense to run all the checks before throwing the exception?
> > I mean run all the checks that the test performs and accumulate failures if any. If the list of failures isn't empty, throw an exception to fail the entire test.
> 
> That might turn out to be useful for debugging a failure, I think.

Yes, it will provide the results for all the points that are verified as well as the screenshot. The screenshot should probably be taken with `createMultiResolutionScreenCapture` especially where the uiScale is above 1. I partially implemented it in my own branch while I was running the tests.

Shall I submit a new issue?

mkartashev commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/HiDPIRobotScreenCaptureTest.java:
```diff
@@ -39,14 +38,12 @@
- * @bug 8073320 8280861
- * @summary  Linux and Windows HiDPI support
+ * @bug 8073320
+ * @summary  Windows HiDPI support
```
> The test is run on Linux, so it makes sense to keep Linux in the summary.
True; will rectify that.

> Why do you remove 8280861 from the list of bugs?
While the test technically executes the lines of code changed for 8280861, I cannot imagine a scenario where this test would fail without the 8280861 fix and wouldn't fail with this fix.

mkartashev commented on test/jdk/java/awt/Robot/HiDPIScreenCapture/HiDPIRobotScreenCaptureTest.java:
```diff
@@ -66,14 +63,7 @@ public static void main(String[] args) throws Exception {
-        // by the system because that area's occupied by the global
-        // menu bar and such.
-        frame.setBounds(83, 97, 400, 300);
+        frame.setBounds(83, 97, 100, 100);
```
Yes, I guess so. Will reinstate the comment.

aivanov-jdk commented:
Test results are green.
(I hope I didn't mess anything up.)

mkartashev commented:
/integrate

aivanov-jdk commented:
Let's give a while for Sergey and/or Phil to take a look. If no comments are received by this evening in Europe, I'll sponsor it.

mkartashev commented:
@aivanov-jdk ping

aivanov-jdk commented:
/sponsor

mkartashev commented:
@aivanov-jdk No problem. Crossing fingers for a successful test run.
